packages/flutter_tools/test/general.shard/macos/native_assets_test.dart

```diff
@@ -0,0 +1,318 @@
+      ProcessManager: () => FakeProcessManager.any(),
+    }, () async {
+      final File packageConfig =
+          environment.projectDir.childFile('.dart_tool/package_config.json');
+      await packageConfig.parent.create();
+      await packageConfig.create();
+      final (Uri? nativeAssetsYaml, _) = await buildNativeAssetsMacOS(
+        darwinArchs: <DarwinArch>[DarwinArch.arm64],
+        projectUri: projectUri,
+        buildMode: BuildMode.debug,
+        fileSystem: fileSystem,
+        flutterTester: flutterTester,
+        buildRunner: FakeNativeAssetsBuildRunner(
+          packagesWithNativeAssetsResult: <Package>[
+            Package('bar', projectUri),
+          ],
+          buildResult: FakeNativeAssetsBuilderResult(
+            assets: <Asset>[
+              Asset(
+                id: 'package:bar/bar.dart',
+                linkMode: LinkMode.dynamic,
+                target: native_assets_cli.Target.macOSArm64,
+                path: AssetAbsolutePath(Uri.file('bar.dylib')),
+              ),
+            ],
+          ),
+        ),
+      );
+      expect(
+        nativeAssetsYaml,
+        projectUri.resolve('build/native_assets/macos/native_assets.yaml'),
+      );
+      expect(
+        await fileSystem.file(nativeAssetsYaml).readAsString(),
+        stringContainsInOrder(<String>[
+          'package:bar/bar.dart',
+          if (flutterTester)
+            // Tests run on host system, so the have the full path on the system.
+            '- ${projectUri.resolve('/build/native_assets/macos/bar.dylib').toFilePath()}'
+          else
+            // Apps are a bundle with the dylibs on their dlopen path.
+            '- bar.dylib',
+        ]),
+      );
+    });
+  }
+
+  // This logic is mocked in the other tests to avoid having test order
+  // randomization causing issues with what processes are invoked.
+  // Exercise the parsing of the process output in this separate test.
+  testUsingContext('NativeAssetsBuildRunnerImpl.cCompilerConfig',
+      overrides: <Type, Generator>{
+        FeatureFlags: () => TestFeatureFlags(isNativeAssetsEnabled: true),
+        ProcessManager: () => FakeProcessManager.list(
+              <FakeCommand>[
+                const FakeCommand(
+                  command: <Pattern>['xcrun', 'clang', '--version'],
+                  stdout: '''
+Apple clang version 14.0.0 (clang-1400.0.29.202)
+Target: arm64-apple-darwin22.6.0
+Thread model: posix
+InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin''',
+                )
+              ],
+            ),
+      }, () async {
+    final NativeAssetsBuildRunner runner =
+        NativeAssetsBuildRunnerImpl(projectUri, fileSystem, logger);
+    final CCompilerConfig result = await runner.cCompilerConfig;
+    expect(
+      result.cc,
+      Uri.file(
+        '/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang',
+      ),
+    );
+  });
+}
```

Comment:
You should add a test to ensure `ensureNoLinkModeStatic` is being used as expected.